src/components/controller.cpp

```diff
@@ -80,42 +80,37 @@ controller::controller(connection& conn, signal_emitter& emitter, const logger&
 
   m_log.trace("controller: Setup user-defined modules");
   size_t created_modules{0};
+  created_modules += setup_modules(alignment::LEFT);
+  created_modules += setup_modules(alignment::CENTER);
+  created_modules += setup_modules(alignment::RIGHT);
 
-  for (int i = 0; i < 3; i++) {
-    alignment align{static_cast<alignment>(i + 1)};
-    string configured_modules;
+  if (!created_modules) {
+    throw application_error("No modules created");
+  }
 
-    if (align == alignment::LEFT) {
-      configured_modules = m_conf.get(m_conf.section(), "modules-left", ""s);
-    } else if (align == alignment::CENTER) {
-      configured_modules = m_conf.get(m_conf.section(), "modules-center", ""s);
-    } else if (align == alignment::RIGHT) {
-      configured_modules = m_conf.get(m_conf.section(), "modules-right", ""s);
-    }
+  /*
+   * Check if each module name only appears once
+   */
+  std::sort(m_modules.begin(), m_modules.end(), [](const auto& m1, const auto& m2) { return m1->name() < m2->name(); });
 
-    for (auto& module_name : string_util::split(configured_modules, ' ')) {
-      if (module_name.empty()) {
-        continue;
-      }
+  auto dup_it = m_modules.cbegin();
 
-      try {
-        auto type = m_conf.get("module/" + module_name, "type");
+  do {
+    auto equal_predicate = [](const auto& m1, const auto& m2) { return m1->name() == m2->name(); };
+    dup_it = std::adjacent_find(dup_it, m_modules.cend(), equal_predicate);
 
-        if (type == "custom/ipc" && !m_ipc) {
-          throw application_error("Inter-process messaging needs to be enabled");
-        }
+    if (dup_it != m_modules.cend()) {
+      m_log.err(
+          "The module \"%s\" appears multiple times in your modules list. "
+          "This is deprecated and should be avoided, as it can lead to inconsistent behavior. "
+          "Both modules will be displayed for now.",
+          (*dup_it)->name());
 
-        m_modules[align].emplace_back(make_module(move(type), m_bar->settings(), module_name, m_log));
-        created_modules++;
-      } catch (const runtime_error& err) {
-        m_log.err("Disabling module \"%s\" (reason: %s)", module_name, err.what());
-      }
+      dup_it = std::find_if_not(dup_it, m_modules.cend(), std::bind(equal_predicate, *dup_it, std::placeholders::_1));
```

Comment:
Maybe a lambda is better but I'm not really sure.

I don't know at which point the compiler is able to optimize the `std::bind` but technically the object returned by `std::bind` can use polymorphism. A lambda is a anonymous struct and so doesn't use polymorphism.

Reply:
It really doesn't matter this is only called once on startup and maybe for a dozen modules.